providers/okta/application.go

```diff
@@ -0,0 +1,252 @@
+package okta
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+	"strings"
+
+	"golang.org/x/net/html"
+	"golang.org/x/net/publicsuffix"
+)
+
+// StateToken is a token extracted from the body of an Okta response.
+//
+// It is used to track a single session.
+type StateToken string
+
+type SignedToken struct {
+	Tx  string
+	App string
+}
+
+// ConcatenateAuthSignature applies the auth signature to the token, returning a completed challenge response that can be submitted to Okta.
+func (tok SignedToken) ConcatenateAuthSignature(authSignature string) string {
+	return strings.Join([]string{authSignature, tok.App}, ":")
+}
+
+// ParseSignedToken extracts a transaction and application signature from a signed token from Okta.
+//
+// A signed token is returned from Okta when initiating MFA with Duo.
+func ParseSignedToken(token string) (SignedToken, error) {
+	parts := strings.Split(token, ":")
+	if len(parts) != 2 {
+		return SignedToken{}, errors.New("wrong number of parts in signed token")
+	}
+
+	return SignedToken{Tx: parts[0], App: parts[1]}, nil
+}
+
+type MultiFactorUpgradeMethod interface {
+	Upgrade(ctx context.Context, client *http.Client) (StateToken, error)
+}
+
+type AuthenticatorEnrollment struct {
+	Key            string
+	ContextualData struct {
+		Host        string
+		SignedToken string
+	}
+}
+
+type Remediation struct {
+	Name   string
+	Type   string
+	Href   string
+	Method string
+}
+
+type IdentifyResponse struct {
+	StateHandle                    string
+	CurrentAuthenticatorEnrollment struct {
+		Type  string
+		Value AuthenticatorEnrollment
+	}
+
+	Remediation struct {
+		Type  string
+		Value []Remediation
+	}
+}
+
+func DetermineUpgradePath(resp IdentifyResponse) (MultiFactorUpgradeMethod, bool) {
+	// Two iterations is fine, this is a small list
+	// Prefer Duo OIDC flow
```

Comment:
I am not sure I understand why we need two loops here. Could you please provide some details? And maybe capture it in a comment here.

Reply:
Sure. We have no guarantees about the order of elements in `resp.Remediation.Value`. We always want to default to OIDC if possible. Looping through the array twice, first checking for OIDC and then for challenge-authenticator is the easiest way to ensure we always prioritize OIDC over challenge-authenticator.